Fixes #502 demo mode to obscure identifying information

jonespm opened the pull request.

jonespm commented on assets/src/containers/DashboardAppBar.js:
```diff
@@ -71,7 +72,7 @@ function DashboardAppBar (props) {
               <Link
                 to={{ pathname: `/courses/${courseId}` }}
-                className={classes.homeButton}>
+                className={clsx(classes.homeButton, user.user_defaults.demomode === "True" ? 'sensitive' : '')}>
```
I thought I'd be using clsx more places. This was used in the material ui documentation and seemed pretty useful.

jonespm commented on dashboard/models.py:
```diff
@@ -49,53 +49,31 @@ class Meta:
             raise e
 
-
-class UserDefaultManager(models.Manager):
```
This manager didn't seem to be needed so I removed it

jonespm commented:
The API for v1/courses/<course_id>/get_user_default_selection

Doesn't return the same format anymore. Before it was {"default": "value"} now it's {"type": "value"} since it actually can return multiple values. This is why some of this was changed. Passing a type is actually optional now and it will return everything for that user within that course context too, which would allow for multiple user settings on a page to be retrieved in one call!

justin0022 commented:
Hi @jonespm, I tried pulling this PR and running it, but am having trouble getting the grade distribution to display. The course name is blurred out after clicking demo mode, but none of the views load correctly. Getting the following error message:
```
object.js:5 Uncaught TypeError: Cannot convert undefined or null to object
    at Function.keys (<anonymous>)
    at isObjectEmpty (object.js:5)
    at eval (GradeDistribution.js:103)
    at commitHookEffectList (react-dom.development.js:3248)
    at commitPassiveHookEffects (react-dom.development.js:3248)
    at HTMLUnknownElement.callCallback (react-dom.development.js:66)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:86)
    at invokeGuardedCallback (react-dom.development.js:102)
    at commitPassiveEffects (react-dom.development.js:3462)
    at wrapped (scheduler-tracing.development.js:177)
isObjectEmpty @ object.js:5
(anonymous) @ GradeDistribution.js:103
commitHookEffectList @ react-dom.development.js:3248
commitPassiveHookEffects @ react-dom.development.js:3248
callCallback @ react-dom.development.js:66
invokeGuardedCallbackDev @ react-dom.development.js:86
invokeGuardedCallback @ react-dom.development.js:102
commitPassiveEffects @ react-dom.development.js:3462
wrapped @ scheduler-tracing.development.js:177
flushPassiveEffects @ react-dom.development.js:3466
dispatchAction @ react-dom.development.js:2550
(anonymous) @ useFetch.js:46
Promise.then (async)
(anonymous) @ useFetch.js:44
commitHookEffectList @ react-dom.development.js:3248
commitPassiveHookEffects @ react-dom.development.js:3248
callCallback @ react-dom.development.js:66
invokeGuardedCallbackDev @ react-dom.development.js:86
invokeGuardedCallback @ react-dom.development.js:102
commitPassiveEffects @ react-dom.development.js:3462
wrapped @ scheduler-tracing.development.js:177
flushFirstCallback @ scheduler.development.js:99
flushWork @ scheduler.development.js:216
channel.port1.onmessage @ scheduler.development.js:618
react-dom.development.js:3237 The above error occurred in the <GradeDistribution> component:
    in GradeDistribution (created by WithStyles(GradeDistribution))
    in WithStyles(GradeDistribution) (at Course.js:42)
    in Route (at Course.js:41)
    in Course (at App.js:41)
    in App (created by Route)
    in Route (created by withRouter(App))
    in withRouter(App) (at src/index.js:10)
    in Router (created by BrowserRouter)
    in BrowserRouter (at src/index.js:9)

Consider adding an error boundary to your tree to customize error handling behavior.
Visit https://fb.me/react-error-boundaries to learn more about error boundaries.
logCapturedError @ react-dom.development.js:3237
logError @ react-dom.development.js:3237
update.callback @ react-dom.development.js:3345
callCallback @ react-dom.development.js:3090
commitUpdateEffects @ react-dom.development.js:3097
commitUpdateQueue @ react-dom.development.js:3097
commitLifeCycles @ react-dom.development.js:3257
commitAllLifeCycles @ react-dom.development.js:3461
callCallback @ react-dom.development.js:66
invokeGuardedCallbackDev @ react-dom.development.js:86
invokeGuardedCallback @ react-dom.development.js:102
commitRoot @ react-dom.development.js:3493
(anonymous) @ react-dom.development.js:3805
unstable_runWithPriority @ scheduler.development.js:255
completeRoot @ react-dom.development.js:3805
performWorkOnRoot @ react-dom.development.js:3786
performWork @ react-dom.development.js:3767
performSyncWork @ react-dom.development.js:3765
requestWork @ react-dom.development.js:3741
commitPassiveEffects @ react-dom.development.js:3463
wrapped @ scheduler-tracing.development.js:177
flushPassiveEffects @ react-dom.development.js:3466
dispatchAction @ react-dom.development.js:2550
(anonymous) @ useFetch.js:46
Promise.then (async)
(anonymous) @ useFetch.js:44
commitHookEffectList @ react-dom.development.js:3248
commitPassiveHookEffects @ react-dom.development.js:3248
callCallback @ react-dom.development.js:66
invokeGuardedCallbackDev @ react-dom.development.js:86
invokeGuardedCallback @ react-dom.development.js:102
commitPassiveEffects @ react-dom.development.js:3462
wrapped @ scheduler-tracing.development.js:177
flushFirstCallback @ scheduler.development.js:99
flushWork @ scheduler.development.js:216
channel.port1.onmessage @ scheduler.development.js:618
react-dom.development.js:180 Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.
    in GradeDistribution (created by WithStyles(GradeDistribution))
    in WithStyles(GradeDistribution) (at Course.js:42)
    in Route (at Course.js:41)
    in Course (at App.js:41)
    in App (created by Route)
    in Route (created by withRouter(App))
    in withRouter(App) (at src/index.js:10)
    in Router (created by BrowserRouter)
    in BrowserRouter (at src/index.js:9)
```

jonespm commented:
Hmm, it looks like the code that is triggering that is
`    at eval (GradeDistribution.js:103)`

I didn't change that code. Are there any errors in the system logs
`docker logs student_dashboard -f`

Or in the console? Maybe the API call was failing?

jonespm commented:
Ok, I added a new PR that fixes the error with no defaults.

Looking at the 2 pages with defaults (resources and grade) I think we should make them consistent and add the Snackbar to the resources too. Maybe that should just be refactored out of the Grades page and inserted at a higher level so it can be triggered from more actions? I can create a new issue about this.

justin0022 commented:
> Ok, I added a new commit that fixes the error with no defaults.
Works now on my end!

> Looking at the 2 pages with defaults (resources and grade) I think we should make them consistent and add the Snackbar to the resources too. Maybe that should just be refactored out of the Grades page and inserted at a higher level so it can be triggered from more actions? I can create a new issue about this.
This PR does exactly that: https://github.com/tl-its-umich-edu/my-learning-analytics/pull/626. After this gets merged, I'm going to refactor the other views with user settings to make it consistent.

jonespm commented:
That's great! I noticed the assignment view wasn't working, but also assignment and resources could be improved to use useSetUserSetting rather than setting it locally there too! Thanks for checking this out.

jonespm commented:
There might be some more areas we need to blue but I think this is fine for now.

jonespm commented:
@justin0022 also commented

> Justin 2:42 PM
> I think a similar approach will work for demo mode. In App.js, you can perform an API call to get the user setting for demo mode, and then if the user clicks the demo mode button, PUT their setting and immediately change the UI.
> exactly, just a local state variable in App.js should work, and the demo state can be passed along to all the views so that the views can hide sensitive info

We could continue to improve on this.

justin0022 commented:
@jonespm send me a ping when this PR is ready for review

pushyamig commented:
In order to make the assignment title blur this is what I tried in the `TableAssigment.js`

```
if(user.user_defaults.demomode === "True"){
              borderAndCurrentWeekStyle['filter'] = 'blur(4px)'
            }
```
in this piece of code


```
 // styling for assignment title column
          if (j === 2) {
            borderAndCurrentWeekStyle['width'] = '30%'
            borderAndCurrentWeekStyle['maxWidth'] = 0
            borderAndCurrentWeekStyle['overflow'] = 'hidden'
            borderAndCurrentWeekStyle['textOverflow'] = 'ellipsis'
            borderAndCurrentWeekStyle['whiteSpace'] = 'nowrap'
            borderAndCurrentWeekStyle['borderRight'] = '0.1em solid rgba(224, 224, 224, 1)'
            if(user.user_defaults.demomode === "True"){
              borderAndCurrentWeekStyle['filter'] = 'blur(4px)'
            }
          }
```

<img width="1355" alt="Screen Shot 2019-08-07 at 11 50 39 AM" src="https://user-images.githubusercontent.com/8579775/62637555-aecc9c80-b909-11e9-8290-5fd9df316fd2.png">

pushyamig commented:
I think you might need to go to `CreateProgressbar.js` and `CreateResourceAccessChart.js` to start obscuring the content with assignment titles and files name. I did not have time today to look at it further. 

There might be an efficient way of doing it than getting into the weed of the files for obscuring the content

jonespm commented:
Yeah, I was trying to ask @jennlove-um or @johnpj if there was more content they wanted obscured. I think if we just had this in there first we could add additional content later. We could probably also improve on how it's doing the hiding. There's just a lot of changes and this is conflicting against most commits. ;)

jonespm commented:
Okay, I did some minor updates again to update this with the changes in the current code. The big change here was I modifed Table.js to allow passing through classNames which it would take and apply. Before it was not using these classes because of the custom table class.

pushyamig commented:
When you are logging as student I see this error 
```
Traceback:  

File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py" in inner
  41.             response = get_response(request)

File "/usr/local/lib/python3.6/site-packages/django/core/handlers/base.py" in _legacy_get_response
  249.             response = self._get_response(request)

File "/usr/local/lib/python3.6/site-packages/django/core/handlers/base.py" in _get_response
  187.                 response = self.process_exception_by_middleware(e, request)

File "/usr/local/lib/python3.6/site-packages/django/core/handlers/base.py" in _get_response
  185.                 response = wrapped_callback(request, *callback_args, **callback_kwargs)

File "/usr/local/lib/python3.6/site-packages/django/contrib/auth/decorators.py" in _wrapped_view
  23.                 return view_func(request, *args, **kwargs)

File "/usr/local/lib/python3.6/site-packages/rules/contrib/views.py" in _wrapped_view
  132.             if not user.has_perms(perms, obj):

File "/usr/local/lib/python3.6/site-packages/django/contrib/auth/models.py" in has_perms
  285.         return all(self.has_perm(perm, obj) for perm in perm_list)

File "/usr/local/lib/python3.6/site-packages/django/contrib/auth/models.py" in <genexpr>
  285.         return all(self.has_perm(perm, obj) for perm in perm_list)

File "/usr/local/lib/python3.6/site-packages/django/contrib/auth/models.py" in has_perm
  277.         return _user_has_perm(self, perm, obj)

File "/usr/local/lib/python3.6/site-packages/django/contrib/auth/models.py" in _user_has_perm
  190.             if backend.has_perm(user, perm, obj):

File "/usr/local/lib/python3.6/site-packages/rules/permissions.py" in has_perm
  32.         return has_perm(perm, user, *args, **kwargs)

File "/usr/local/lib/python3.6/site-packages/rules/permissions.py" in has_perm
  24.     return permissions.test_rule(name, *args, **kwargs)

File "/usr/local/lib/python3.6/site-packages/rules/rulesets.py" in test_rule
  6.         return name in self and self[name].test(*args, **kwargs)

File "/usr/local/lib/python3.6/site-packages/rules/predicates.py" in test
  154.             return bool(self._apply(*args))

File "/usr/local/lib/python3.6/site-packages/rules/predicates.py" in _apply
  213.         result = self.fn(*callargs)

File "/usr/local/lib/python3.6/site-packages/rules/predicates.py" in OR
  165.             return self._combine(other, operator.or_, args)

File "/usr/local/lib/python3.6/site-packages/rules/predicates.py" in _combine
  194.         other_result = other._apply(*args)

File "/usr/local/lib/python3.6/site-packages/rules/predicates.py" in _apply
  213.         result = self.fn(*callargs)

File "/code/dashboard/rules.py" in is_enrolled_in_course
  15.         return User.objects.get_user_in_course(user, course).count()

File "/usr/local/lib/python3.6/site-packages/django/db/models/manager.py" in manager_method
  85.                 return getattr(self.get_queryset(), name)(*args, **kwargs)

File "/code/dashboard/models.py" in get_user_in_course
  274.             course_id=course.id
```
This is not an issue when logged in a Admin.

jonespm commented on dashboard/settings.py:
```diff
@@ -130,6 +130,7 @@
                 'dashboard.context_processors.current_user_courses_info',
                 'dashboard.context_processors.last_updated',
                 'dashboard.context_processors.get_build_info',
+                'dashboard.context_processors.get_user_defaults',
```
Yeah, in the normal process of browsing it will only do it once. I felt this was better than constantly going out to get it. As far as database calls are concerned I think we'll have to figure out the best way to cache those but that's a different problem.

jonespm commented on dashboard/context_processors.py:
```diff
@@ -16,4 +17,9 @@ def last_updated(request):
+
+def get_user_defaults(request):
+    user_sis_name = request.user.get_username()
+    defaults = db_util.get_user_defaults(user_sis_name, 0)
```
Yes, all of the same context_processors are hit when using LTI.

jonespm commented:
@pushyamig Thanks for reviewing, I'm looking into this and I can reproduce your error. I'll also look into the assignment title blur too.

jonespm commented:
@pushyamig I believe I've fixed the issue you mentioned. There was a problem with the objectgetter code so I just switched it out. Could you try this again?

jonespm commented:
@pushyamig I added support for blocking the assignment titles in the latest commit. 0d7d68422f0d1c90dd079db07a5f49477a1d5cc1

If you could take a look at that it'd be appreciated. I couldn't seem to easily get just the title to be blocked so I just blurred the whole svg element. I think that's *fine* for this view.

jonespm commented:
Finally caught this up and resolved all the conflicts. This feature really turned out to be a lot more effort that I thought it was going to be.

jonespm commented on dashboard/settings.py:
```diff
@@ -452,6 +453,12 @@ def show_debug_toolbar(request):
+    These have to be updated if you add additional settings
+    Added as a security check to prevent unwanted types being added
+"""
+VALID_DEFAULT_SELECTION_TYPES=['grade','resource','demomode']
```
I fixed this, thanks for noticing

pushyamig commented:
Review: 
1. When demo mode is selected and in assignment view, I am not sure why the entire progress bar needs to be blocked. Not My Decision
I feel both these errors are due to getting/setting defaults from `user_default_Selection` table. The below errors were not happening  before something new 
2. Assignment View: When the user has defaults set from earlier then assignment is throwing an error. 
3. Resource View: When a user trying to set the default, slider disappears and "No data provided" message is put

jonespm commented:
I think I'm just going to close this, there's too many changes still going on in the views and the way this was implemented in the views probably needs to be re-done anyway. When I have a chance I'll split it into two issues, one to "Support demo mode in the UI" and another for the backend changes of "Support setting more than one selection per view and global user settings"
